File: 2022/august/08-08-2022/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums):
        length = len(nums)
        dp = [0] * (length+1)
        for i in range(length):
            curr_max = 0
            for j in range(i):
                if nums[j] < nums[i]:
                    curr_max = max(curr_max, dp[j])
            dp[i] = curr_max+1
        return max(dp)


print(Solution().lengthOfLIS([10,9,2,5,3,7,101,18]))

File: 2022/august/08-08-2022/longest-increasing-subsequence.rs
impl Solution {
    pub fn length_of_lis(mut nums: Vec<i32>) -> i32 {
        let mut ans = 1;
        for i in 1..nums.len() {
            let j = nums[..ans].partition_point(|&x| x < nums[i]);
            nums[j] = nums[i];
            ans = ans.max(j + 1);
        }
        ans as i32
    }
}
